### DISCUSSION TITLE
Is there a tutorial for the STM32WB55 ?

### DISCUSSION THREAD

**Discussion by patfrench** [micropython/micropython#16341] [STM32 / Pyboard]
Hello,
Before buying a "nucleo stm32wb55"

- I would like to know if there is a tutorial to know how to use "rfcore" files ?

- I would like to know if micropython supports the capacitive touch of the stm32wb55?

Thanks

**Comment by davefes** [comment]
Are you wanting to do LoRa?

**Reply by patfrench** [reply]
Hi,
Not lora. I only need BLE (and maybe zigbee).
Thanks

**Comment by andrewleech** [comment]
I don't know of any tutorial as such, there's just a few usage notes in the files themselves. V1.13 is the latest known working version of the rfcore firmware. 

I haven't used capacitive touch, don't know if it works.

**Reply by patfrench** [reply]
Thank you @andrewleech

Can I ask you a technical question?

I have a little script that sends data via the UART. I have to set the RTS line to 1 immediately after sending (I used a TXIDLE interrupt)
This script works perfectly
```python
import machine
import time

# Fonction d'interruption appelée lorsque la transmission est terminée
def tx_complete(irq_uart):
    print("Transmission terminée. RTS = 1")
    time.sleep(0.05)
    rts.value(1)  # Remet RTS à 1 quand la transmission est finie
    

# Initialisation de l'UART et des broches
uart = machine.UART(1, baudrate=1200, tx=machine.Pin(4), rx=machine.Pin(5), parity=1)
uart.init(1200, bits=8, parity=1, stop=1)
uart.irq(handler=tx_complete, trigger=machine.UART.IRQ_TXIDLE)

rts = machine.Pin(6, machine.Pin.OUT)  # RTS en sortie

commande = bytes.fromhex("FFFFFFFFFFFFFFFFFFFF0280000082")
# Configuration de l'interruption pour détecter la fin de la transmission

while True:
    rts.value(0)  # RTS à 0 pour activer l'envoi
    # Envoi des données
    uart.write(commande)

    # Attente de réception pour tester (facultatif)
    time.sleep(0.8)
    data = uart.read(128)
    if data:
        print("Données reçues :", data)
    time.sleep(3)
```
I'd like to stop using an RP2 and switch to STM32 but the micropython documentation says that STM32 only supports IRQ_RXIDLE and IRQ_RX.
How would you do it ? I need to calculate my buffer send time ?

Thanks in advance

**Reply by robert-hh** [reply]
On STM32 uart.write() is not buffered. So after uart.write() all but the last byte should have been sent.

**Reply by patfrench** [reply]
Good evening Robert,

Thank you for your reply 

I will try uart.flush() and/or tx.done() later this week.

I had bad results with uart.flush() on RP2 but I understand it's because of the buffer.

Is there a way to capture the flags
TXEIE and TXFNIE ?
```python
static const machine_uart_irq_map_t mp_uart_irq_map[] = {
    { USART_CR1_IDLEIE, UART_FLAG_IDLE}, // RX idle
    { USART_CR1_PEIE,   UART_FLAG_PE},   // parity error
    #if defined(STM32G0) || defined(STM32WL)
    { USART_CR1_TXEIE_TXFNFIE, UART_FLAG_TXE}, // TX register empty
    #else
    { USART_CR1_TXEIE,  UART_FLAG_TXE},  // TX register empty
    #endif
    { USART_CR1_TCIE,   UART_FLAG_TC},   // TX complete
```
They are in uart.c
Thanks you

**Reply by patfrench** [reply]
😳😳😳😳😳

I just read in the source code that the flag is used here
https://github.com/micropython/micropython/blob/ee8d8b3448f26abafbfe1ed7f8970e02f4f8d232/ports/stm32/uart.c#L1158

micropython is really well done

**Comment by robert-hh** [comment]
uart.flush() for RP2 was changed 8 weeks ago to better match the end of transmission time. It should now wait until the data has been sent.

**Reply by patfrench** [reply]
Thank you Robert for your help

Do you know if it's possible to manipulate microcontroller registers (other than Inline assembler) ?
USART_CR2 register and the RTOIE in the USART_CR1 register for the mode MODBUS/RTU (for example)

Thank you

**Reply by robert-hh** [reply]
You can use the machine.mem32(), .mem16() and .mem8() methods for writing and reading to memory mapped registers. See https://docs.micropython.org/en/latest/library/machine.html#machine.mem32

**Reply by andrewleech** [reply]
On stm32 all the registers are directly exposed by name in the `stm` module: https://docs.micropython.org/en/latest/library/stm.html#peripheral-register-constants

**Reply by patfrench** [reply]
perfect !
Thank you @andrewleech and @robert-hh

**Reply by patfrench** [reply]
Hello @andrewleech and @robert-hh 

According to you, for a commercial development you will select :

- STM32WB55 ?
- STM32 + murata BLE chip ?

There are a same stack BLE ? NimBLE ? Is that correct ?

I'd like to have the best support

Thank you

**Reply by andrewleech** [reply]
The WB will be cheaper and smaller than a separate stm with CYW43439 based radio, however the separated approach also gives wifi and I'd say more reliable support.

The radio firmware for the WB does cause a lot of grief; beside the incompatibility with the current versions it's also seemingly quite easy to break requiring a somewhat complex reflashing procedure.

The separate radio chip should work on any CYW43439 module I think, the cheapest I've found with antenna included is https://www.digikey.com/en/products/detail/ezurio/453-00084R/15850988

The CYW43439 also gives the option to switch to btstack which will be easier to get ble sig qualification for a commercial project, though btstack does require a separate licence for commercial use.

That being said there aren't many examples of it being used that I know of, really just the pyboard-d for the stm port. This is a great reference design, though I don't know if it'll work out of the box with other stm chips. It might though.

**Reply by patfrench** [reply]
Hi @andrewleech 

Thank you for your reply. 
I will take the “separate” solution 

I think I'm confused about the bluetooth stack is this the nimBLE stack ? 

> The CYW43439 also gives the option to switch to btstack which will be easier to get ble sig qualification for a commercial project, though btstack does require a separate licence for commercial use.

nimBLE stack can not be used in commercial projects ?

Thank you

**Reply by andrewleech** [reply]
nimBLE stack certainly can be used for commercial projects, I've done so in the past. The only complication is when it comes to BLE Sig qualification. 

Last time I checked there was no "active" pre-qualification of any components of the nimBLE stack, which it has been through qualification previously those results can only be used for a limited time before it needs to be re-qualified from scratch. I don't know if anyone has done so recently.
I'm not sure if I'm making a lot of sense there... BLE qualification is needlessly complicated and expensive, but if it's not done the sig does make efforts to detect and enforce qualifications for shipped products so it's important to try to do the "right thing" with any "end products" being shipped with BLE declared as a feature.

btstack on the other hand is always pre-qualified by the author / owner I believe, however you have to speak to them to organise an appropriate licencing deal to use it in a commercial product.
